File: plugins/flutter-fleet/mgl_fleet_native_sdk/android/src/main/kotlin/com/mgl/fleet/flutter/MglFleetNativeSdkPlugin.kt
package com.mgl.fleet.flutter

import androidx.fragment.app.FragmentActivity
import com.mgl.fleet.sdk.FleetSdk
import com.mgl.fleet.sdk.FleetSdkCompletionCallback
import com.mgl.fleet.sdk.FleetSdkOptions
import com.mgl.fleet.sdk.FleetSdkResult
import com.mgl.fleet.sdk.FleetSessionOptions
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class MglFleetNativeSdkPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var activity: FragmentActivity? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "mgl_fleet_native_sdk")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity as? FragmentActivity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity as? FragmentActivity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> handleInitialize(call, result)
            "isInitialized" -> result.success(FleetSdk.isInitialized())
            "presentFleetFlow" -> handlePresent(call, result)
            "openFleetNativeFlow" -> handleOpenFleetNativeFlow(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handleInitialize(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *> ?: run {
            result.error("BAD_ARGS", "Expected argument map", null)
            return
        }
        val ctx = activity?.applicationContext ?: run {
            result.error("NO_ACTIVITY", "Attach Flutter engine to an Activity before initialize", null)
            return
        }
        val opts = parseFleetOptions(args) ?: run {
            result.error("BAD_ARGS", "apiBaseUrl required", null)
            return
        }
        FleetSdk.initialize(ctx, opts)
        result.success(null)
    }

    private fun handleOpenFleetNativeFlow(call: MethodCall, result: MethodChannel.Result) {
        val root = call.arguments as? Map<*, *> ?: run {
            result.error("BAD_ARGS", "Expected argument map", null)
            return
        }
        val initMap = root["initialize"] as? Map<*, *> ?: run {
            result.error("BAD_ARGS", "initialize object is required", null)
            return
        }
        val ctx = activity?.applicationContext ?: run {
            result.error("NO_ACTIVITY", "Attach Flutter engine to an Activity before openFleetNativeFlow", null)
            return
        }
        val opts = parseFleetOptions(initMap) ?: run {
            result.error("BAD_ARGS", "apiBaseUrl required", null)
            return
        }
        FleetSdk.initialize(ctx, opts)

        val presentMap = root["present"] as? Map<*, *>
        val corr = presentMap?.get("correlationId") as? String
        executePresent(corr, result)
    }

    private fun handlePresent(call: MethodCall, result: MethodChannel.Result) {
        if (!FleetSdk.isInitialized()) {
            result.error(
                "NOT_INITIALIZED",
                "Fleet SDK is not initialized. Use openFleetNativeFlow() or await initialize() before presentFleetFlow().",
                null,
            )
            return
        }
        val args = call.arguments as? Map<*, *> ?: emptyMap<String, Any?>()
        val corr = args["correlationId"] as? String
        executePresent(corr, result)
    }

    private fun executePresent(correlationId: String?, result: MethodChannel.Result) {
        val act = activity ?: run {
            result.error("NO_ACTIVITY", "Need FragmentActivity (FlutterFragmentActivity)", null)
            return
        }
        FleetSdk.presentFleetFlow(
            act,
            correlationId?.let { FleetSessionOptions(correlationId = it) },
            object : FleetSdkCompletionCallback {
                override fun onComplete(sdk: FleetSdkResult) {
                    act.runOnUiThread {
                        when (sdk) {
                            is FleetSdkResult.Success -> {
                                val payloadMap = HashMap<String, Any?>()
                                sdk.payload.forEach { (k, v) -> payloadMap[k] = v }
                                result.success(
                                    mapOf(
                                        "event" to sdk.event,
                                        "payload" to payloadMap,
                                    ),
                                )
                            }

                            is FleetSdkResult.Failure -> {
                                result.error(
                                    sdk.exception.code.toString(),
                                    sdk.exception.message,
                                    null,
                                )
                            }
                        }
                    }
                }
            },
        )
    }

    private fun parseFleetOptions(args: Map<*, *>): FleetSdkOptions? {
        val base = (args["apiBaseUrl"] as? String)?.trim() ?: return null
        if (base.isEmpty()) return null
        val useMock = args["useMock"] as? Boolean ?: true
        val token = args["authToken"] as? String
        val foCompanyId = parseFoCompanyId(args["foCompanyId"])
        return FleetSdkOptions(
            apiBaseUrl = base,
            authToken = token,
            useMock = useMock,
            foCompanyId = foCompanyId,
        )
    }

    private fun parseFoCompanyId(raw: Any?): Long? {
        return when (raw) {
            null -> null
            is Long -> raw
            is Int -> raw.toLong()
            is Double -> raw.toLong()
            is Number -> raw.toLong()
            is String -> raw.trim().toLongOrNull()
            else -> null
        }
    }
}

File: plugins/flutter-fleet/mgl_fleet_native_sdk/ios/Classes/MglFleetNativeSdkPlugin.swift
import Flutter
import UIKit

#if canImport(MGLFleetSDK)
import MGLFleetSDK
#endif

public class MglFleetNativeSdkPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "mgl_fleet_native_sdk", binaryMessenger: registrar.messenger())
    let instance = MglFleetNativeSdkPlugin()
    channel.setMethodCallHandler(instance.handle(_:result:))
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      handleInitialize(call, result: result)
    case "isInitialized":
      #if canImport(MGLFleetSDK)
      result(FleetSdk.shared.isInitialized())
      #else
      result(FlutterError(code: "NO_NATIVE_SDK", message: Self.noSdkMessage, details: nil))
      #endif
    case "presentFleetFlow":
      handlePresent(call, result: result)
    case "openFleetNativeFlow":
      handleOpenFleetNativeFlow(call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  #if canImport(MGLFleetSDK)
  private func handleInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let apiBaseUrl = (args["apiBaseUrl"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
          !apiBaseUrl.isEmpty else {
      result(FlutterError(code: "BAD_ARGS", message: "apiBaseUrl required", details: nil))
      return
    }
    let useMock = args["useMock"] as? Bool ?? true
    let authToken = args["authToken"] as? String
    let foCompanyId = Self.readFoCompanyId(fromAny: args["foCompanyId"])
    FleetSdk.shared.initialize(
      options: FleetSdkOptions(
        apiBaseUrl: apiBaseUrl,
        authToken: authToken,
        useMock: useMock,
        foCompanyId: foCompanyId,
      ),
    )
    result(nil)
  }

  private func handleOpenFleetNativeFlow(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let root = call.arguments as? [String: Any],
          let initBlock = root["initialize"] as? [String: Any],
          let apiBaseUrl = (initBlock["apiBaseUrl"] as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines),
          !apiBaseUrl.isEmpty else {
      result(FlutterError(code: "BAD_ARGS", message: "initialize.apiBaseUrl required", details: nil))
      return
    }
    let useMock = initBlock["useMock"] as? Bool ?? true
    let authToken = initBlock["authToken"] as? String
    let foCompanyId = Self.readFoCompanyId(fromAny: initBlock["foCompanyId"])
    FleetSdk.shared.initialize(
      options: FleetSdkOptions(
        apiBaseUrl: apiBaseUrl,
        authToken: authToken,
        useMock: useMock,
        foCompanyId: foCompanyId,
      ),
    )
    let present = root["present"] as? [String: Any]
    executePresent(presentArgs: present, result: result)
  }

  private func handlePresent(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard FleetSdk.shared.isInitialized() else {
      result(
        FlutterError(
          code: "NOT_INITIALIZED",
          message:
            "Fleet SDK is not initialized. Use openFleetNativeFlow() or await initialize() before presentFleetFlow().",
          details: nil,
        ),
      )
      return
    }
    let args = call.arguments as? [String: Any]
    executePresent(presentArgs: args, result: result)
  }

  private func executePresent(presentArgs: [String: Any]?, result: @escaping FlutterResult) {
    guard let presenter = Self.presentingViewController() else {
      result(FlutterError(code: "NO_ACTIVITY", message: "No root view controller", details: nil))
      return
    }
    let correlationId = presentArgs?["correlationId"] as? String
    let session = correlationId.map { FleetSessionOptions(correlationId: $0) }

    do {
      try FleetSdk.shared.presentFleetFlow(from: presenter, session: session) { sdkResult in
        DispatchQueue.main.async {
          switch sdkResult {
          case let .success(event, payload):
            result(["event": event, "payload": payload])
          case let .failure(err):
            result(FlutterError(code: String(err.code.rawValue), message: err.message, details: nil))
          }
        }
      }
    } catch let err as FleetSdkError {
      result(FlutterError(code: String(err.code.rawValue), message: err.message, details: nil))
    } catch {
      result(FlutterError(code: "1004", message: error.localizedDescription, details: nil))
    }
  }
  #endif

  #if !canImport(MGLFleetSDK)
  private func handleInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    result(FlutterError(code: "NO_NATIVE_SDK", message: Self.noSdkMessage, details: nil))
  }

  private func handleOpenFleetNativeFlow(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    result(FlutterError(code: "NO_NATIVE_SDK", message: Self.noSdkMessage, details: nil))
  }

  private func handlePresent(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    result(FlutterError(code: "NO_NATIVE_SDK", message: Self.noSdkMessage, details: nil))
  }
  #endif

  private static let noSdkMessage =
    "Link MGLFleetSDK to the iOS Runner target (SPM or CocoaPods). See plugins/flutter-fleet/mgl_fleet_native_sdk/README.md"

  private static func readFoCompanyId(fromAny raw: Any?) -> Int64? {
    guard let raw else { return nil }
    if let i = raw as? Int64 { return i }
    if let i = raw as? Int { return Int64(i) }
    if let n = raw as? NSNumber { return n.int64Value }
    if let d = raw as? Double { return Int64(d) }
    if let s = raw as? String { return Int64(s.trimmingCharacters(in: .whitespacesAndNewlines)) }
    return nil
  }

  private static func presentingViewController() -> UIViewController? {
    guard let root = rootViewController() else { return nil }
    return topPresented(from: root)
  }

  private static func rootViewController() -> UIViewController? {
    guard let scene = UIApplication.shared.connectedScenes.first as? UIWindowScene else { return nil }
    let window = scene.windows.first { $0.isKeyWindow } ?? scene.windows.first
    return window?.rootViewController
  }

  private static func topPresented(from root: UIViewController) -> UIViewController {
    var top = root
    while let presented = top.presentedViewController {
      top = presented
    }
    return top
  }
}
